### PR TITLE
feat:handle version confilct

### PR DESCRIPTION
**Pull request by fedeoo**
Consider the condition that there are different version modules, Only one of them should be resolved correctly.
If  we are failed to add the peerDependencies of module A , we shouldn't handle A as a external module.

### CODE REVIEW

**Comment by mastilver** [comment]
Hi @fedeoo 

Thank you for your PR! :)

Is this really an issue? `urls` is an object, so if there is a module is required twice, it will be overwritten (Which I agree is not the best solution)

Would you mind adding some tests

**Comment by fedeoo** [comment]
@mastilver  I will add some tests soon.   
I know urls is an object,  but just for a case：A project depends React@15.3.1 and a UI module named A, A depend on React@0.14.  we will store both https://unpkg.com/react@15.3.1/dist/react.min.js and https://unpkg.com/react@^0.14/dist/react.min.js to urls. This may be lead some exceptions.

**Comment by mastilver** [comment]
No it will store only one, the latest one

The real issue is that it will use the wrong version for one of the dependencies

Can you also get rid of your code around peerDependencies

We also need to address the case when:
A@1 -> B@1
C@1 -> B@2 (peerDependency)
B@2 (because of C)

A can't be served through cdn, B and C can

- A@1 is required: it's being bundled
- B@1 is required by A: it's being served from the cdn
- C@1 is required: it's being served from the cdn / It will use B@1 as it's exposing a global var
- B@2 is required: it's being bundled / it will not be use as C is already using B@1

I'm not sure how to fix that yet, feel free to just summit failing tests if you are not sure how to resolve that

**Comment by fedeoo** [comment]
May be it's not a case, Actually we just get a warning when we install those packages. 
I was wrong about the peerDependency. I will close this PR. Thanks for your remand.

**Comment by fedeoo** [comment]
There is still a case we should deal with. 
When we are failed to resolve B from CDN, A and C shouldn't be served from the CDN(in most case will be failed as well).

**Comment by mastilver** [comment]
I still think this is an issue, thank you for making me realise it #22

**Comment by mastilver** [comment]
> When we are failed to resolve B from CDN, A and C shouldn't be served from the CDN(in most case will be failed as well).

This should be dealt by `module-to-cdn`, there is an issue for that: https://github.com/mastilver/module-to-cdn/issues/5 ;)

**Comment by mastilver** [comment]
@fedeoo Damn, I just realise your approach with `peerDependencies` was good, I will put some comments on this PR, if you want to work on that

If you don't that's fine, I will try to spend some time on it next week :)

File: src/index.js [hunk 1]
```diff
@@ -81,21 +82,33 @@ export default class ModulesCdnWebpackPlugin {
...
+            return false;
+        }
+
         if (this.verbose) {
             console.log(`✔️ '${cdnConfig.name}' will be served by ${cdnConfig.url}`);
         }
 
+        let result = cdnConfig.var;
         if (peerDependencies) {
             for (const peerDependencyName in peerDependencies) {
```
Comment (mastilver) [inline src/index.js]: I would prefer something like that:

```js
let arePeerDependenciesLoaded = true;
             for (const peerDependencyName in peerDependencies) {
                  if ({}.hasOwnProperty.call(peerDependencies, peerDependencyName)) {
                     const isLoaded = Boolean(this.addModule(contextPath, peerDependencyName, {env}));
                     arePeerDependenciesLoaded = arePeerDependenciesLoaded && isLoaded;
                  }
              }

              if (!arePeerDependenciesLoaded) {
                  return false;
              }
```

**Comment by mastilver** [comment]
@fedeoo fixed that on https://github.com/mastilver/modules-cdn-webpack-plugin/commit/7257b5ffd12a3213077c096f51e77f6a1742ae56

**Comment by fedeoo** [comment]
@mastilver 👍

**Comment by mastilver** [comment]
Finally release: https://github.com/mastilver/dynamic-cdn-webpack-plugin/releases/tag/v3.3.0

NOTE: the name changed, I hated it... I think it's clearer now!

**Pull request closed without merging**